homeassistant/components/mqtt/__init__.py

```diff
@@ -375,26 +405,50 @@ def __init__(self, hass, broker, port, client_id, keepalive, username,
         self._mqttc.on_connect = self._mqtt_on_connect
         self._mqttc.on_disconnect = self._mqtt_on_disconnect
         self._mqttc.on_message = self._mqtt_on_message
+
         if will_message:
             self._mqttc.will_set(will_message.get(ATTR_TOPIC),
                                  will_message.get(ATTR_PAYLOAD),
                                  will_message.get(ATTR_QOS),
                                  will_message.get(ATTR_RETAIN))
-        self._mqttc.connect(broker, port, keepalive)
+
+        self._mqttc.connect_async(broker, port, keepalive)
 
     def publish(self, topic, payload, qos, retain):
```

Comment:
Our instance is only ever used by functions in this component. No other platform will touch it. We should be able to remove the sync methods.